activesupport/lib/active_support/core_ext/class/attribute.rb

```diff
@@ -96,14 +96,16 @@ def class_attribute(*attrs, instance_accessor: true,
       namespaced_name = :"__class_attr_#{name}"
       ::ActiveSupport::ClassAttribute.redefine(self, name, namespaced_name, default)
 
-      delegators = [
-        "def #{name}; #{namespaced_name}; end",
-        "def #{name}=(value); self.#{namespaced_name} = value; end",
-      ]
+      class_methods << "def #{name}; #{namespaced_name}; end"
+      class_methods << "def #{name}=(value); self.#{namespaced_name} = value; end"
 
-      class_methods.concat(delegators)
       if singleton_class?
-        methods.concat(delegators)
+        methods << <<~RUBY if instance_reader
+          silence_redefinition_of_method(:#{name})
+          def #{name}
+            self.singleton_class.#{name}
+          end
+        RUBY
```

Comment:
The writer that was previously defined here shouldn't be needed. It is correctly set below when `instance_writer` is set. Also, previously, the reader was defined unconditionally when it should be checking `instance_reader`.

Note however that this reader definition preserves some confusing behaviour (see https://github.com/rails/rails/issues/55776#issuecomment-3341321483).